app/uk/gov/hmrc/gform/validation/ValidationService.scala

```diff
@@ -619,32 +589,119 @@ class ComponentsValidator(
   def validateConcreteDate(
     fieldValue: FormComponent,
     localDate: LocalDate,
-    concreteDate: LocalDate,
-    offset: OffsetDate,
-    dateError: GformError)(func: (LocalDate, LocalDate, OffsetDate) => Boolean): ValidatedType =
+    concreteDate: ConcreteDate,
+    offset: OffsetDate = OffsetDate(0),
+    dateError: GformError)(func: (LocalDate, ConcreteDate, OffsetDate) => Boolean): ValidatedType =
     func(localDate, concreteDate, offset) match {
       case true  => ().valid
       case false => dateError.invalid
     }
 
+  def concreteDateFunctionMatch(beforeAfterPrecisely: BeforeAfterPrecisely)(
+    date: LocalDate,
+    concreteDate: ConcreteDate,
+    offset: OffsetDate): Boolean = {
+    val parametersLength = concreteDate.getExactParameters.length
+    beforeAfterPrecisely match {
+      case Before if concreteDate.isExact => isBeforeConcreteDate(date, concreteDate, offset)
+
+      case After if concreteDate.isExact => isAfterConcreteDate(date, concreteDate, offset)
+
+      case Precisely =>
+        if (concreteDate.isExact) {
```

Comment:
Aim to make each case a one-liner, delegating to another method/function as necessary.